### ISSUE TITLE
[Hold #20949] [$1000] The "Thanks for flagging this" whisper message disappears from the LHN preview when deleting an earlier flagged message

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#20633]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Action Performed:
1. Log-in to newDot using accountA
2. Send two messages to accountB i.e "Test 1" and "Test 2"
3. Log-in to newDot using accountB in an incognito browser
4. Hover over the message from accountA > click the flag > flag as inappropriate
5. Observe the LHN preview message displays a preview of the whisper moderation message
6. Switch over to accountA and delete one of the message sent
7. Switch back over to accountB and observe the LHN preview message now displays "Test 2"

## Expected Result: The "Thanks for flagging this" m
The whisper message should still be displayed in the LHN preview row after deleting an earlier message in the chat history

## Actual Result: 
When deleting the earlier flagged message the whisper message disappears from the LHN preview. 

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.26-3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/f1936364-01fc-47ad-b3d9-0e4c1f217564

https://github.com/Expensify/App/assets/43996225/a02380ff-0e59-4296-9add-9c8357c11766

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ayazhussain79 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1686150650431619

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ad862ad486daa2be</li>
        <li>Upwork Job ID: 1669435614628663296</li>
        <li>Last Price Increase: 2023-07-06</li>
    </ul>
</details>

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The "Thanks for flagging this" whisper message disappears from the LHN preview when deleting an earlier flagged message.

### What is the root cause of that problem?

The root cause of this issue is that the backend returns invalid `lastMessageText` data for `DeleteComment` API call. 

### What changes do you think we should make in order to solve the problem?

I confirmed that the frontend set right optimistic data of `lastMessageText` here

https://github.com/Expensify/App/blob/7b00c47ee7debf7137417ac768efe8d4c6b872d0/src/libs/actions/Report.js#L813

To fix this issue, we should have fix from backend to set right `lastMessageText` for `DeleteComment` API call. 

After the backend fix, we can check this issue again.

### What alternative solutions did you explore? (Optional)

N/A

**Comment by Victor-Nyagudi** [comment]
I found this bug also occurs if rather than switching to a different account and deleting a message (step 6), you open a different report and switch back to the report you flagged the message in while on the same account.

The `OpenReport` API call returns the wrong `lastMessageText` value, so I'm with @eh2077 on this one for both root cause and solution.

**Comment by abdulrahuman5196** [comment]
Will create a slack thread to discuss this. Similar issue - https://github.com/Expensify/App/issues/20613#issuecomment-1596937086

**Comment by abdulrahuman5196** [comment]
@trjExpensify  Created a thread here to discuss if we can solve similar issues related to LHN subtitle - https://expensify.slack.com/archives/C01GTK53T8Q/p1687190693298329

**Comment by trjExpensify** [comment]
Okay, did you land in a good spot with @iwiznia?

**Comment by abdulrahuman5196** [comment]
Not confirmed yet since there where less participation on the thread. I have pinged again on the slack.

**Comment by trjExpensify** [comment]
Seems like we have a good summary, so let's move on with next steps @abdulrahuman5196.

**Comment by abdulrahuman5196** [comment]
@trjExpensify  I am able to reproduce this issue easily in a different way.

1) Flag any comment in a chat A and observe the "Only visible to you" message comes up.
2) Observe the LHN it shows the "Thanks for flagging this comment for the"
3) Go to any other chat and come back to the chat A.
4) Observe the LHN changes to the last actual text message instead of "Thanks for flagging this comment for the"


https://github.com/Expensify/App/assets/46707890/df8759d3-7c83-4927-b92b-73765268951b




I checked the network response and in server response the last message is only set to the actual text message not the "Thanks for flagging this comment for the".
My hypothesis is somewhere locally we setting this "Thanks for flagging this comment for the" and when server response comes back it changes.


But the question now is, What is our expectation of what should be shown in the LHN for this case? Should we show the "Thanks for flagging this comment for the" content or the last message content? 
**Based on this expectation only we can direct this issue in a certain path of fix**

<img width="1512" alt="Screenshot 2023-06-29 at 8 04 27 PM" src="https://github.com/Expensify/App/assets/46707890/787dd3c1-568f-438e-ab30-c10da3471942">

**Comment by trjExpensify** [comment]
>But the question now is, What is our expectation of what should be shown in the LHN for this case? Should we show the "Thanks for flagging this comment for the" content or the last message content?
Based on this expectation only we can direct this issue in a certain path of fix

Okay, I think the question you're asking here is whether a whisper message should show in the LHN preview for those that can see it? 

If the message marks the chat as unread (i.e the chat row becomes bold and notifies) for the user, which I think it does, then yes it should show in the LHN preview. @JmillsExpensify @iwiznia do you agree with that?

**Comment by iwiznia** [comment]
Hmmmmm IIRC we did that on purpose... @pecanoro do you remember?

**Comment by abdulrahuman5196** [comment]
>If the message marks the chat as unread (i.e the chat row becomes bold and notifies) for the user, which I think it does, then yes it should show in the LHN preview

Just to note. Room welcome whisper messages doesn't seem to show up in LHN. I think backend doesn't consider any whisper message on LHN subtitle(Just assumption could be wrong).

**Comment by pecanoro** [comment]
Whispers message should not show on the LHN at any point. If the flagged whisper was showing there initially is because we are setting it optimistically. That's actually the bug that needs to be fixed 😄

**Comment by trjExpensify** [comment]
>Whispers message should not show on the LHN at any point.

Gooot it, do they also not mark the chat as unread for the receiver of it?

**Comment by abdulrahuman5196** [comment]
> >Whispers message should not show on the LHN at any point.
> 
> Gooot it, do they also not mark the chat as unread for the receiver of it? 

I think so @trjExpensify . Correct me if wrong @pecanoro

**Comment by abdulrahuman5196** [comment]
@eh2077 @Victor-Nyagudi  Could you kindly update yours proposals based on the latest discussion.

**Comment by trjExpensify** [comment]
We have a week before this breaks WAQ, it would be great to get a proposal agreed upon and into PR review early this week. @eh2077 are you still interested?

**Comment by eh2077** [comment]
@trjExpensify I just checked the issue again and found the root cause is same as mentioned previously https://github.com/Expensify/App/issues/20633#issuecomment-1593971798. I think we'll still need the backend fix to return right `lastMessageText` for flagging API request.

cc @abdulrahuman5196

**Comment by abdulrahuman5196** [comment]
@eh2077 We shouldn't show the whisper in the lastMessageText. AFAIK backend is not doing that, we are only optimistically setting it now and it should be removed.

**Comment by trjExpensify** [comment]
>@eh2077 We shouldn't show the whisper in the lastMessageText. AFAIK backend is not doing that, we are only optimistically setting it now and it should be removed.

Do you want to update your proposal with this in mind? If you can do that today, please do, otherwise I'll re-pull the trigger to get an internal engineer assigned to take this over given that it's +4 weeks old and now breaking WAQ. 👍 CC: @pecanoro maybe one you're interested in! :)

**Comment by eh2077** [comment]
@trjExpensify I tried to get this issue fixed from frontend only but I can't figure out the solution as expected by @abdulrahuman5196 . I still think it's a backend based on the RCA.

**Comment by abdulrahuman5196** [comment]
@eh2077 Could you provide information on why you think its still a backend issue?

The current requirement is to remove the optimistic addition of whisper messages to lastMessageText, 

But in your proposal you had mentioned

>I confirmed that the frontend set right optimistic data of lastMessageText here

Shouldn't this be removed in that case?

**Comment by eh2077** [comment]
> @eh2077 Could you provide information on why you think its still a backend issue?
> 
> The current requirement is to remove the optimistic addition of whisper messages to lastMessageText,
> 
> But in your proposal you had mentioned
> 
> > I confirmed that the frontend set right optimistic data of lastMessageText here
> 
> Shouldn't this be removed in that case?

@abdulrahuman5196 
Ah, sorry I overlooked this case. Yes, we should fix the optimistic data when deleting a comment. I updated [proposal](https://github.com/Expensify/App/issues/20633#issuecomment-1593971798) to fix it. Below video reproduces the bug in offline mode.

https://github.com/Expensify/App/assets/117511920/302d2834-7e90-4451-b11d-b5f94c65a917

At the meanwhile, I think the backend fix is also necessary, because when flagging a comment from account A
1. LHN message changes to whisper message after API call 
https://github.com/Expensify/App/blob/451096b4828cf9a9bc1b5b7be95a6d52e79c392d/src/libs/actions/Report.js#L1855
2. From account B, pusher event also updates LHN message to whisper message

See below video to reproduce the backend issue.


https://github.com/Expensify/App/assets/117511920/70c39514-b298-40a7-af7e-b8a15ae27a69

**Comment by eh2077** [comment]
Just went through the dupe steps again, I think the backend issue I mentioned could be a separate issue though they're both related to showing whisper message on LHN.

@trjExpensify Should we have a separate bug report for the backend issue mentioned above? Or solve them together in this issue?

cc @pecanoro @abdulrahuman5196

**Comment by trjExpensify** [comment]
Good question, @pecanoro would you be up for taking the backend issue here if you agree with it?

**Comment by abdulrahuman5196** [comment]
@Nodebrute Thank your for the information. I think its partially correct, will provide more context in next comment.

**Comment by abdulrahuman5196** [comment]
> Whispers message should not show on the LHN at any point. If the flagged whisper was showing there initially is because we are setting it optimistically. That's actually the bug that needs to be fixed 😄

@pecanoro I checked this and it doesn't seem the value was set optimistically.

`FlagComment` api response is setting the lastMessageText to the whisper message, it is followed by `OpenReport` API call which fixes the issue. That's why we see the whisper in LHN and it goes away again. 

**So `FlagComment` should be fixed in backend.**

Meanwhile we also have frontend issues.

1) We have filter to remove whisper messages before showing it in the LHN here. But it is not working on this case for some reason.

https://github.com/Expensify/App/blob/451096b4828cf9a9bc1b5b7be95a6d52e79c392d/src/libs/OptionsListUtils.js#L396-L407


I think that blocking the message before showing up on LHN was made as quick fix to solve some other issue. **But IMO its just hiding the issues behind and we should remove it and make sure whisper messages are not circulating in the report data.**
For example, during `DeleteComment` we optimistically set whisper as a lastMessageText, which should be fixed.



@trjExpensify Given this, we would backend fix and frontend fixes as an improvement. Can we assign an internal engineer here to work on backend fix and oversee frontend proposal approvals/path forward.

**Comment by trjExpensify** [comment]
>@trjExpensify Given this, we would backend fix and frontend fixes as an improvement. Can we assign an internal engineer here to work on backend fix and oversee frontend proposal approvals/path forward.

Yup, I haven't heard back from Rocio, so let me re-pull the engineering trigger to have it assigned out!

**Comment by pecanoro** [comment]
The flagging system was implemented by @dangrous so he might have a better idea why this is being returned when it should not! We adapted the queries not to return whispers but I am not sure if you added a new query or it's being returned by some other reaons.

**Comment by dangrous** [comment]
oh so we had a bug previously that you couldn't flag a (whispered) welcome message from another user, so we modified part of FlagComment to allow whispers... But I'm pretty sure that was only going to Concierge on the backend - not ever being sent to the front end. It does look like the query for getting the `mostRecentVisibleAction` from FlagComment was modified recently (in [this PR](https://github.com/Expensify/Auth/pull/8159)), could that have something to do with it? cc @aldo-expensify

**Comment by aldo-expensify** [comment]
My understanding is that the whispers are not shown in the LHN because the last message LHN is the same for everyone and whispers are not for everyone.

This may be related to this issue which is currently in progress: https://github.com/Expensify/App/issues/22386

Related PRs:

- Auth: https://github.com/Expensify/Auth/pull/8170
- Web: https://github.com/Expensify/Web-Expensify/pull/37910
- App: https://github.com/Expensify/App/pull/20949

If those PR's get merged deployed, maybe this issue will go away? 🤷

**Comment by pecanoro** [comment]
Ahh yeah, that is what I mentioned, this is probably not coming from the back-end if it's disappearing after the action finishes but contributors mentioned it was coming from the back-end so I was super confused.

**Comment by aldo-expensify** [comment]
Since the backend PRs took a while to review/merge, the App https://github.com/Expensify/App/pull/20949 PR is now full of conflicts. Once the Web PR gets deployed, I'll fix conflicts and retest, then I could also have a look at this case too and see if there is anything else to do.

**Comment by abdulrahuman5196** [comment]
We are waiting on above PRs to be merged

**Comment by aldo-expensify** [comment]
I did some work on fixing the conflicts for https://github.com/Expensify/App/pull/20949, but I still have to retest everything

**Comment by trjExpensify** [comment]
Thanks, @aldo-expensify. Are you okay to take Chirag's assignment here as he's on parental leave now? 

In the meantime, I'm going to pop a hold in the title for #20949. After which we can retest to confirm whether we have more work to do to fix this case.

**Comment by aldo-expensify** [comment]
I haven't circled back to retesting yet, I'm a bit caught with higher priority issues. I'll try to come back to this later during the week.

**Comment by trjExpensify** [comment]
#20949 is still in draft, any luck with moving it forward Aldo?

**Comment by trjExpensify** [comment]
^^ bump on the above @aldo-expensify

**Comment by trjExpensify** [comment]
@aldo-expensify, [bump](https://github.com/Expensify/App/issues/20633#issuecomment-1667631119) on this please! :)

**Comment by aldo-expensify** [comment]
Sorry, I had completely forgotten about this issue, I'll added it to my notes to get back to it asap

**Comment by aldo-expensify** [comment]
So, I started investigating this today. The expected outcome is wrong as far as I know. Whispers are never supposed to go to the LHN. See this conversation: https://expensify.slack.com/archives/C04JD99283A/p1691618167413869?thread_ts=1691528393.777399&cid=C04JD99283A

Now, in the steps described it is mentioned that the whisper message appears in the LHN after flagging the message, if this is true, that is a bug to fix.

Working on reproducing...

**Comment by aldo-expensify** [comment]
Retested and this is not happening anymore

### Whisper after flagging as harrasment doesn't make it to LHN:

<img width="1657" alt="Screenshot 2023-09-01 at 4 03 03 PM" src="https://github.com/Expensify/App/assets/87341702/a99ff4c6-e454-4a00-bcf3-01df8c737dab">

https://github.com/Expensify/App/assets/87341702/08e1395a-a3d9-4a9b-b6db-71faa62655e1

### The correct message and not a whisper is shown in the LHN after deleting messages

https://github.com/Expensify/App/assets/87341702/8cbdb6d9-4e2e-40c5-8610-0cea9195a83a

### Deleting a message while offline shows the right thing in the LHN
https://github.com/Expensify/App/assets/87341702/67cbe30d-d4d4-4994-be88-2ec5848ac1fe

**Comment by ayazhussain79** [comment]
@trjExpensify @aldo-expensify am i eligible for reporting bonus?
